lib/src/main/java/macrobase/datamodel/DataFrame.java

```diff
@@ -0,0 +1,311 @@
+        return select(this.schema.getColumnIndices(columns));
+    }
+
+    /**
+     * @param mask rows to select
+     * @return new dataframe with subset of rows
+     */
+    public DataFrame filter(boolean[] mask) {
+        DataFrame other = new DataFrame();
+
+        int d = schema.getNumColumns();
+        int numTrue = 0;
+        for (int i = 0; i < numRows; i++) {
+            if (mask[i]) {
+                numTrue++;
+            }
+        }
+        for (int c = 0; c < d; c++) {
+            ColType t = schema.getColumnType(c);
+            String columnName = schema.getColumnName(c);
+            if (t == ColType.STRING) {
+                String[] oldColumn = getStringColumn(c);
+                String[] newColumn = new String[numTrue];
+                int j = 0;
+                for (int i = 0; i < numRows; i++) {
+                    if (mask[i]) {
+                        newColumn[j] = oldColumn[i];
+                        j++;
+                    }
+                }
+                other.addStringColumn(columnName, newColumn);
+            } else if (t == ColType.DOUBLE) {
+                double[] oldColumn = getDoubleColumn(c);
+                double[] newColumn = new double[numTrue];
+                int j = 0;
+                for (int i = 0; i < numRows; i++) {
+                    if (mask[i]) {
+                        newColumn[j] = oldColumn[i];
+                        j++;
+                    }
+                }
+                other.addDoubleColumn(columnName, newColumn);
+            } else {
+                throw new RuntimeException("Bad Column Type");
+            }
+        }
+        return other;
+    }
+    public DataFrame filterString(int columnIdx, Predicate<String> filter) {
+        String[] filterColumn = getStringColumn(columnIdx);
+        boolean[] mask = new boolean[numRows];
+        for (int i = 0; i < numRows; i++) {
+            mask[i] = filter.test(filterColumn[i]);
+        }
+        return filter(mask);
+    }
+    public DataFrame filterStringByName(String columnName, Predicate<String> filter) {
+        return filterString(schema.getColumnIndex(columnName), filter);
+    }
+
+    /**
+     * @param columnIdx column index to filter by
+     * @param filter predicate to test each column value
+     * @return new dataframe with subset of rows
+     */
+    public DataFrame filterDouble(int columnIdx, DoublePredicate filter) {
+        double[] filterColumn = getDoubleColumn(columnIdx);
+        boolean[] mask = new boolean[numRows];
+        for (int i = 0; i < numRows; i++) {
+            mask[i] = filter.test(filterColumn[i]);
+        }
+        return filter(mask);
+    }
+
+    /**
+     * @param columnName column name to filter by
+     * @param filter predicate to test each column value
+     * @return new dataframe with subset of rows
+     */
+    public DataFrame filterDoubleByName(String columnName, DoublePredicate filter) {
```

Comment:
can't you just call this `filter` and let the compiler figure out whether you're calling with `DoublePredicate` or `Predicate<String>`?